src/Nethermind/Nethermind.Baseline/BaselineTree.cs

```diff
@@ -0,0 +1,253 @@
+using System;
+using System.Runtime.CompilerServices;
+using Nethermind.Core.Crypto;
+using Nethermind.Serialization.Rlp;
+using Nethermind.Trie;
+
+[assembly: InternalsVisibleTo("Nethermind.Baseline.Test")]
+
+namespace Nethermind.Baseline
+{
+    public abstract partial class BaselineTree
+    {
+        private const int LeafRow = 32;
+        private const int LeafLevel = 0;
+        public const int TreeHeight = 32;
+        private const ulong FirstLeafIndexAsNodeIndex = MaxNodes / 2;
+        private const ulong MaxNodes = (1ul << (TreeHeight + 1)) - 1ul;
+        private const ulong MaxNodeIndex = MaxNodes - 1;
+
+        private readonly IKeyValueStore _keyValueStore;
+        private readonly byte[] _dbPrefix;
+        private protected int TruncationLength { get; }
+
+        /* baseline does not use a sparse merkle tree - instead they use a single zero hash value
+           does it expose any attack vectors? */
+        internal static Bytes32 ZeroHash = Bytes32.Zero;
+
+        public uint Count { get; private set; }
+
+        public BaselineTree(IKeyValueStore keyValueStore, byte[] dbPrefix, int truncationLength = 0)
+        {
+            _keyValueStore = keyValueStore ?? throw new ArgumentNullException(nameof(keyValueStore));
+            _dbPrefix = dbPrefix ?? throw new ArgumentNullException(nameof(dbPrefix));
+            TruncationLength = truncationLength;
+            Count = LoadCount();
+        }
+
+        private uint LoadCount()
+        {
+            // this is an incorrect binary search approach
+            // that will fail if any of the leaves are zero hashes
+            // we should read count from the corresponding contract
+            
+            ulong left = GetMinNodeIndex(LeafRow);
+            ulong right = GetMaxNodeIndex(LeafRow);
+            ulong? topIndex = Binary.Search(left, right, ni => !ZeroHash.Equals(LoadValue(ni)));
+            if (!topIndex.HasValue)
+            {
+                return 0;
+            }
+            
+            return GetLeafIndex(topIndex.Value) + 1;
+        }
+        
+        private byte[] BuildDbKey(ulong nodeIndex)
+        {
+            return Rlp.Encode(Rlp.Encode(_dbPrefix), Rlp.Encode(nodeIndex)).Bytes;
```

Comment:
Make DB keys simpler - just uint - can be kept sorted in DB easier

Reply:
you either prefix it and store it in the same database or remove the prefix and keep the dbs separte
if we use the db format that we discussed then we no longer need a common DB